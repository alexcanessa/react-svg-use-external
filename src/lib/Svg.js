/* @flow */

import * as React from "react";

type Props = {
  children?: React.Node,
  forwardedRef?: React.Ref<"svg">,
  [key: string]: any
};

type State = {
  viewBox: ?string
};

export const Context = React.createContext();

class Svg extends React.Component<Props, State> {
  state = {
    viewBox: undefined
  };

  _setViewBox(viewBox) {
    this.setState({ viewBox });
  }

  _context = {
    setViewBox: this._setViewBox.bind(this)
  };

  _href = null;

  render() {
    const { children, viewBox, forwardedRef, ...props } = this.props;
    return (
      <svg
        viewBox={viewBox || this.state.viewBox}
        ref={forwardedRef}
        {...props}
      >
        <Context.Provider value={this._context}>{children}</Context.Provider>
      </svg>
    );
  }
}

export default React.forwardRef((props, ref) => (
  <Svg {...props} forwardedRef={ref} />
));
