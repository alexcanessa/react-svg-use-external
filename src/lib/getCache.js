const CACHE_KEY = "__SVG_USE_DOCUMENT_CACHE__";

const cacheNamespace =
  typeof document === "object" ? document : /* istanbul ignore next */ {};

export default function getCache() {
  if (!cacheNamespace[CACHE_KEY]) {
    cacheNamespace[CACHE_KEY] = new Map();
  }
  return cacheNamespace[CACHE_KEY];
}

export function resetCache() {
  delete cacheNamespace[CACHE_KEY];
}
