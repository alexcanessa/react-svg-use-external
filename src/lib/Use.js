/* @flow */

import * as React from "react";
import { Context } from "./Svg";
import getCache from "./getCache";
import url from "url";

type Props = {
  xlinkHref?: string,
  href?: string,
  forwardedRef?: React.Ref<"use" | "g">,
  [key: string]: any
};

type State = {
  loaded: boolean,
  content: ?Array<Node>
};

function getHref(props: Props): ?string {
  return props.href || props.xlinkHref;
}

class Use extends React.Component<Props, State> {
  state = {
    loaded: false
  };

  _href = null;

  async _load(href: ?string) {
    if (!href || href[0] === "#") {
      // Our fallback to <use> will do in these cases
      this.setState({ loaded: false, content: undefined });
      return;
    }
    if (this._href === href && this.state.loaded) {
      return;
    }
    this._href = href;
    const hrefUrl = url.parse(
      url.resolve(typeof window === "object" ? window.location.href : "", href)
    );
    const hash = hrefUrl.hash;
    hrefUrl.hash = "";
    const resource = url.format(hrefUrl);

    this.setState({
      loaded: false,
      content: undefined // Hygiene
    });

    const windowOrigin =
      window.location.origin ||
      // IE < 11
      window.location.protocol +
        "//" +
        window.location.hostname +
        (window.location.port ? ":" + window.location.port : "");

    const origin = `${hrefUrl.protocol}//${hrefUrl.host}`;

    if (typeof window === "object" && windowOrigin !== origin) {
      console.error(
        `[react-svg-use-external] Refusing to fetch cross-origin SVG ${href} from origin ${windowOrigin}. This is to align with browsers' native behaviour.`
      );
      return;
    }

    const cache = getCache();

    let parsedDocument = cache.get(resource);
    if (!parsedDocument) {
      try {
        parsedDocument = document.implementation.createHTMLDocument("");
        parsedDocument.body.innerHTML = await (await fetch(resource)).text();
        if (parsedDocument.domain !== document.domain) {
          parsedDocument.domain = document.domain;
        }
      } catch (e) {
        return; // state is either { loaded: false } or managed by a later request
      }
      cache.set(resource, parsedDocument);
    }

    if (this._href !== href) {
      return;
    }

    const template = parsedDocument.getElementById(hash.substr(1));
    if (template) {
      // TODO: Recursively flatten <use> references in `template`
      // • Do this _before_ cloning, so results can be shared as long as the document is in the cache.
      // • Do this for both local (in `parsedDocument`) and remote references.
      // • Make sure to break circular references, in line with the SVG spec.
      // • Generalise most of the code that currently makes up `Use._load`, as it is directly applicable
      //   to resolving external references from a DOM context as well.
      const {
        svgContext: { setViewBox }
      } = this.props;

      const viewBox = template.getAttribute("viewBox");
      if (viewBox) {
        setViewBox(viewBox);
      }

      const clone = template.cloneNode(true);
      const contentNodes = [];
      while (clone.childNodes.length) {
        contentNodes.push(clone.removeChild(clone.firstChild));
      }

      this.setState({ loaded: true, content: contentNodes });
    } else {
      this.setState({ loaded: true, content: undefined });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const prevHref = getHref(prevProps);
    const href = getHref(this.props);
    if (href !== prevHref) {
      this._load(href);
    }
    const element = this._element.current;
    if (this.state.loaded && element) {
      const { content } = this.state;
      if (!content || content[0] !== element.firstChild) {
        while (element.firstChild) {
          element.removeChild(element.firstChild);
        }
        if (content) {
          for (const node of content) {
            element.appendChild(node);
          }
        }
      }
    }
  }

  componentDidMount() {
    this._load(getHref(this.props));
  }

  _element = React.createRef();

  render() {
    const { forwardedRef, svgContext, xlinkHref, href, ...props } = this.props;
    const { loaded } = this.state;
    return loaded ? (
      forwardedRef ? (
        // Can't both forward and use a ref, so wrap with another element.
        <g ref={forwardedRef} {...props}>
          <g ref={this._element} />
        </g>
      ) : (
        <g ref={this._element} {...props} />
      )
    ) : (
      <use ref={forwardedRef} xlinkHref={xlinkHref} href={href} {...props} />
    );
  }
}

export default React.forwardRef((props, ref) => (
  <Context.Consumer>
    {svgContext => (
      <Use {...props} forwardedRef={ref} svgContext={svgContext} />
    )}
  </Context.Consumer>
));
