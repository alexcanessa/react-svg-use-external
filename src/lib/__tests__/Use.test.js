import React from "react";
import ReactDOMServer from "react-dom/server";
import Use from "../Use";
import Svg from "../Svg";
import { render } from "react-testing-library";
import fetchMock, { MATCHED, UNMATCHED } from "fetch-mock";
import { resetCache } from "../getCache";

beforeEach(resetCache);

fetchMock.config.warnOnFallback = false;

describe("Use polyfill", () => {
  describe("without external URL", () => {
    test("should render a <use> element", () => {
      const { container } = render(<Use />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test("renders a local xlink:href", () => {
      const { container } = render(<Use xlinkHref="#foo" />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test("renders a local href", () => {
      const { container } = render(<Use href="#foo" />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test("renders both types of xlink:href", () => {
      const { container } = render(<Use xlinkHref="#foo" />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should pass through attributes", () => {
      const { container } = render(<Use data-foo="bar" viewBox="0 0 1 1" />);
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should expose a <use> ref", () => {
      const ref = React.createRef();

      render(<Use ref={ref} />);
      expect(ref.current).toBeInstanceOf(Element);
      expect(ref.current.tagName).toMatch(/^use$/i);
    });
  });
  describe("with external URL", () => {
    // NOTE: This assumes Jest's testURL is set to the default of "http://localhost"
    beforeEach(() => {
      fetchMock.getOnce(
        (url, opts) =>
          url === "http://localhost/sprites.svg" && opts.mode === "same-origin",
        {
          headers: {
            "content-type": "image/svg+xml"
          },
          body: ReactDOMServer.renderToStaticMarkup(
            <svg xmlns="http://www.w3.org/2000/svg" style={{ display: "none" }}>
              <symbol viewBox="0 0 20 20" id="symbol-1">
                <circle cx="10" cy="10" r="10" stroke="black" fill="red" />
              </symbol>
              <symbol id="symbol-2">
                <ellipse
                  cx="10"
                  cy="10"
                  rx="10"
                  ry="5"
                  stroke="black"
                  fill="red"
                />
              </symbol>
            </svg>
          )
        }
      );
      fetchMock.getOnce(
        (url, opts) =>
          url === "http://localhost/sprites-slow.svg" &&
          opts.mode === "same-origin",
        async () => {
          await new Promise(resolve => setTimeout(resolve, 200));
          return {
            headers: {
              "content-type": "image/svg+xml"
            },
            body: ReactDOMServer.renderToStaticMarkup(
              <svg
                xmlns="http://www.w3.org/2000/svg"
                style={{ display: "none" }}
              >
                <symbol viewBox="0 0 20 20" id="symbol-slow-1">
                  <circle cx="10" cy="10" r="10" stroke="green" fill="yellow" />
                </symbol>
              </svg>
            )
          };
        }
      );
      fetchMock.getOnce(
        (url, opts) =>
          url === "http://localhost/sprites-404.svg" &&
          opts.mode === "same-origin",
        404
      );
    });
    afterEach(() => {
      fetchMock.reset();
    });
    test("should request the external resource via href", async () => {
      render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.called(MATCHED)).toBe(true);
    });
    test("should request the external resource via xlinkHref", async () => {
      render(
        <Svg>
          <Use xlinkHref="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.called(MATCHED)).toBe(true);
    });
    test("should prefer the href prop when requesting a resource", async () => {
      render(
        <Svg>
          <Use
            href="http://localhost/sprites.svg#symbol-1"
            xlinkHref="http://localhost/sprites-nope.svg#symbol-1"
          />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.called(MATCHED)).toBe(true);
      expect(fetchMock.called(UNMATCHED)).toBe(false);
    });
    test("should not request a cross-origin resource", async () => {
      render(
        <Svg>
          <Use href="http://www.example.com/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.called(MATCHED)).toBe(false);
    });
    describe("emulating missing location.origin", () => {
      // IE 11 doesn't support location.origin
      beforeAll(() => {
        const originProp = Object.getOwnPropertyDescriptor(
          window.location,
          "origin"
        );
        Object.defineProperty(window.location, "origin", {
          configurable: true,
          value: undefined
        });
      });
      afterAll(() => {
        delete window.location.origin;
        if (originProp) {
          Object.defineProperty(window.location, "origin", originProp);
        }
      });
      test("should not request a cross-origin resource", async () => {
        render(
          <Svg>
            <Use href="http://www.example.com/sprites.svg#symbol-1" />
          </Svg>
        );
        await fetchMock.flush();
        expect(fetchMock.called(MATCHED)).toBe(false);
      });
    });
    test.skip("should cache the external resource across synchronous renders", async () => {
      // FIXME: This inadvertently hits an edge case where rerendering before the response is fully
      // handled causes an unintended cache miss. TODO: Rewrite internals to use DataLoader
      // instead.
      const { rerender } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      rerender(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-nope" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.calls(MATCHED)).toHaveLength(1);
    });
    test("should cache the external resource across renders", async () => {
      const { rerender } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve();
      rerender(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-nope" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.calls(MATCHED)).toHaveLength(1);
    });
    test("should render a plain <use> tag while loading", () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should embed the referenced symbol in the output", async () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should pass props through to the embedded symbol", async () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" data-foo="bar" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should expose a <g> ref once loaded", async () => {
      const ref = React.createRef();
      render(
        <Svg>
          <Use ref={ref} href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(ref.current).toBeInstanceOf(Element);
      expect(ref.current.tagName).toMatch(/^g$/i);
    });
    test("should output <g /> if the referenced symbol doesn't exist", async () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-nope" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should fall back to <use /> if the request hits a network error", async () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites-nope.svg" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should fall back to <use /> if the request hits a 404", async () => {
      const { container } = render(
        <Svg>
          <Use href="http://localhost/sprites-404.svg" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("should not change if rerendering with the same href", async () => {
      const { container, rerender } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      await Promise.resolve(); // Let components handle response
      const reference = container.cloneNode(true);
      rerender(
        <Svg>
          <Use
            // Change the prop from href to xlinkHref to force a (shallow) rerender
            xlinkHref="http://localhost/sprites.svg#symbol-1"
          />
        </Svg>
      );
      // Guard against a flash of blank content because of the rerender.
      expect(container).toContainHTML(reference.innerHTML);

      // Guard against an async load getting triggered at all.
      await fetchMock.flush();
      await Promise.resolve();
      expect(fetchMock.called(UNMATCHED)).toBe(false);
      expect(container).toContainHTML(reference.innerHTML);
    });
    test("tracks the latest href despite another fetch completing later", async () => {
      const { container, rerender } = render(
        <Svg>
          <Use href="http://localhost/sprites-slow.svg#symbol-slow-1" />
        </Svg>
      );
      rerender(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.calls(MATCHED)).toHaveLength(2);
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    test("tracks the latest href despite another fetch completing sooner", async () => {
      const { container, rerender } = render(
        <Svg>
          <Use href="http://localhost/sprites.svg#symbol-1" />
        </Svg>
      );
      rerender(
        <Svg>
          <Use href="http://localhost/sprites-slow.svg#symbol-slow-1" />
        </Svg>
      );
      await fetchMock.flush();
      expect(fetchMock.calls(MATCHED)).toHaveLength(2);
      await Promise.resolve(); // Let components handle response
      expect(container.firstChild).toMatchSnapshot();
    });
    describe("viewBox", () => {
      test("should set and clear as appropriate", async () => {
        const { container, rerender } = render(
          <Svg>
            <Use href="http://localhost/sprites.svg#symbol-1" />
          </Svg>
        );
        await fetchMock.flush();
        await Promise.resolve(); // Let components handle response
        expect(container.firstChild).toMatchSnapshot();
        rerender(
          <Svg>
            <Use href="http://localhost/sprites.svg#symbol-2" />
          </Svg>
        );
        await fetchMock.flush();
        await Promise.resolve(); // Let components handle response
        expect(container.firstChild).toMatchSnapshot();
      });
      test("should not override Svg's viewBox", async () => {
        const { container, rerender } = render(
          <Svg viewBox="1 2 3 4">
            <Use href="http://localhost/sprites.svg#symbol-1" />
          </Svg>
        );
        await fetchMock.flush();
        await Promise.resolve(); // Let components handle response
        expect(container.firstChild).toMatchSnapshot();
      });
    });
  });
});
