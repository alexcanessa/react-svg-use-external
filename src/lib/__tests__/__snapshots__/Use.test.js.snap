// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use polyfill with external URL should embed the referenced symbol in the output 1`] = `
<svg
  viewBox="0 0 20 20"
>
  <g>
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL should fall back to <use /> if the request hits a 404 1`] = `
<svg>
  <use
    href="http://localhost/sprites-404.svg"
  />
</svg>
`;

exports[`Use polyfill with external URL should fall back to <use /> if the request hits a network error 1`] = `
<svg>
  <use
    href="http://localhost/sprites-nope.svg"
  />
</svg>
`;

exports[`Use polyfill with external URL should output <g /> if the referenced symbol doesn't exist 1`] = `
<svg>
  <g />
</svg>
`;

exports[`Use polyfill with external URL should pass props through to the embedded symbol 1`] = `
<svg
  viewBox="0 0 20 20"
>
  <g
    data-foo="bar"
  >
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL should render a plain <use> tag while loading 1`] = `
<svg>
  <use
    href="http://localhost/sprites.svg#symbol-1"
  />
</svg>
`;

exports[`Use polyfill with external URL tracks the latest href despite another fetch completing later 1`] = `
<svg
  viewBox="0 0 20 20"
>
  <g>
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL tracks the latest href despite another fetch completing sooner 1`] = `
<svg
  viewBox="0 0 20 20"
>
  <g>
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="green"
      fill="yellow"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL viewBox should not override Svg's viewBox 1`] = `
<svg
  viewBox="1 2 3 4"
>
  <g>
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL viewBox should set and clear as appropriate 1`] = `
<svg
  viewBox="0 0 20 20"
>
  <g>
    <circle
      cx="10"
      cy="10"
      r="10"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill with external URL viewBox should set and clear as appropriate 2`] = `
<svg>
  <g>
    <ellipse
      cx="10"
      cy="10"
      rx="10"
      ry="5"
      stroke="black"
      fill="red"
    />
  </g>
</svg>
`;

exports[`Use polyfill without external URL renders a local href 1`] = `
<use
  href="#foo"
/>
`;

exports[`Use polyfill without external URL renders a local xlink:href 1`] = `
<use
  xlink:href="#foo"
/>
`;

exports[`Use polyfill without external URL renders both types of xlink:href 1`] = `
<use
  xlink:href="#foo"
/>
`;

exports[`Use polyfill without external URL should pass through attributes 1`] = `
<use
  data-foo="bar"
  viewbox="0 0 1 1"
/>
`;

exports[`Use polyfill without external URL should render a <use> element 1`] = `<use />`;
